flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxProcessor.java

```diff
@@ -0,0 +1,319 @@
+		// TODO consider batched draining into list and/or limit number of executed letters
+		// Take letters in a non-blockingly and execute them.
+		Optional<Runnable> maybeLetter;
+		while (isMailboxLoopRunning() && (maybeLetter = mailbox.tryTakeMail()).isPresent()) {
+			maybeLetter.get().run();
+		}
+
+		// If the default action is currently not available, we can run a blocking mailbox execution until the default
+		// action becomes available again.
+		while (isDefaultActionUnavailable() && isMailboxLoopRunning()) {
+			mailbox.takeMail().run();
+		}
+
+		return isMailboxLoopRunning();
+	}
+
+	/**
+	 * Calling this method signals that the mailbox-thread should (temporarily) stop invoking the default action,
+	 * e.g. because there is currently no input available.
+	 */
+	private MailboxDefaultAction.SuspendedDefaultAction suspendDefaultAction() {
+
+		assert taskMailboxExecutor.isMailboxThread();
+
+		if (suspendedDefaultAction == null) {
+			suspendedDefaultAction = new SuspendDefaultActionRunnable();
+			ensureControlFlowSignalCheck();
+		}
+
+		return suspendedDefaultAction;
+	}
+
+	private boolean isDefaultActionUnavailable() {
+		return suspendedDefaultAction != null;
+	}
+
+	private boolean isMailboxLoopRunning() {
+		return mailboxLoopRunning;
+	}
+
+	/**
+	 * Helper method to make sure that the mailbox loop will check the control flow flags in the next iteration.
+	 */
+	private void ensureControlFlowSignalCheck() {
+		// Make sure that mailbox#hasMail is true via a dummy letter so that the flag change is noticed.
+		if (!mailbox.hasMail()) {
+			try {
+				mailbox.tryPutMail(() -> {});
+			} catch (MailboxStateException me) {
+				LOG.debug("Mailbox closed when trying to submit letter for control flow signal.", me);
+			}
+		}
+	}
+
+	/**
+	 * Implementation of {@link MailboxDefaultAction.ActionContext} that is connected to a {@link MailboxProcessor}
+	 * instance.
+	 */
+	private final class MailboxDefaultActionContext implements MailboxDefaultAction.ActionContext {
+
+		private final MailboxProcessor mailboxProcessor;
+
+		private MailboxDefaultActionContext(MailboxProcessor mailboxProcessor) {
+			this.mailboxProcessor = mailboxProcessor;
+		}
+
+		@Override
+		public void allActionsCompleted() {
+			mailboxProcessor.allActionsCompleted();
+		}
+
+		@Override
+		public MailboxDefaultAction.SuspendedDefaultAction suspendDefaultAction() {
+			return mailboxProcessor.suspendDefaultAction();
+		}
+	}
+
+	/**
+	 * Represents the suspended state of the default action and offers an idempotent method to resume execution.
+	 */
+	private final class SuspendDefaultActionRunnable implements MailboxDefaultAction.SuspendedDefaultAction {
+
+		/** Ensuring idempotent behavior, we ensure this is only accessed from the main thread. */
+		private boolean valid;
+
+		SuspendDefaultActionRunnable() {
+			this.valid = true;
+		}
+
+		@Override
+		public void resume() {
+			try {
+				if (taskMailboxExecutor.isMailboxThread()) {
+					resumeInternal();
```

Comment:
I'd move the mailbox thread clause outside of `try`:
```java
if (taskMailboxExecutor.isMailboxThread()) {
    resumeInternal();
    return;
}
try {
    ...
} catch (...) {
    ...
}

Reply:
Makes sense, the method worked a bit different before but now we can limit the scope of the block 👍